modin/core/storage_formats/pandas/query_compiler.py

```diff
@@ -2094,16 +2094,328 @@ def clip(self, lower, upper, **kwargs):
+                )
+
+                # Using `.values` here so we can ignore the indices (it's really hard
+                # to arrange them for pandas to properly perform the summation)
+                parts_with_nans = total_agg.values[:, na_agg_idxs]
+                parts_without_nans = (
+                    total_agg.values[:, non_na_agg_idxs]
+                    # Before doing the summation we have to align the shapes
+                    # Imagine that we have 'parts_with_nans' like:
+                    #    sum   pow2_sum  count
+                    #    a  b  a  b      a  b
+                    # a  1  2  3  4      5  6
+                    # b  1  2  3  4      5  6
+                    #
+                    # And the 'parts_without_nans' like:
+                    #    sum  pow2_sum  count
+                    # a  1    3         5
+                    # b  2    4         6
+                    #
+                    # Here we want to sum them in an order so the digit matches (1 + 1), (2 + 2), ...
+                    # For that we first have to repeat the values in 'parts_without_nans':
+                    #  parts_without_nans.repeat(parts_with_nans.shape[0]):
+                    #    sum  pow2_sum  count
+                    # a  1    3         5
+                    # b  1    3         5
+                    # a  2    4         6
+                    # b  2    4         6
+                    #
+                    # And then reshape it using the "Fortran" order:
+                    #  parts_without_nans.reshape(parts_with_nans.shape, order="F"):
+                    #    sum   pow2_sum  count
+                    #    a  b  a  b      a  b
+                    # a  1  2  3  4      5  6
+                    # b  1  2  3  4      5  6
+                    # After that the shapes & orders are aligned and we can perform the summation
+                    .repeat(repeats=len(parts_with_nans), axis=0).reshape(
+                        parts_with_nans.shape, order="F"
+                    )
+                )
+                replace_values = parts_with_nans + parts_without_nans
+
+                if not total_agg.values.flags.writeable:
+                    # making a copy if the buffer is read-only as
+                    # we will need to modify `total_agg` inplace
+                    total_agg = total_agg.copy()
+                total_agg.values[:, na_agg_idxs] = replace_values
+
+            sum_of_pairwise_mul = total_agg["mul"]
+            sums = total_agg["sum"]
+            sums_of_squares = total_agg["pow2_sum"]
+            count = total_agg["count"]
+
+            cols = sum_of_pairwise_mul.columns
+            # If there were NaNs in the original dataframe then have computed a matrix
+            # of sums/square sums/counts at the Map phase, meaning that we now have multiple
+            # columns in `sums`.
+            has_nans = len(sums.columns) > 1
+            if not has_nans:
```

Comment:
isn't this assumption false for a case when every column/row has a NaN?

Reply:
why? the processing is the same for all the cases where there were NaNs at the source frame, it doesn't matter how many NaN values we met as we would still compute a pairwise matrix of aggregations at the Map phase and so would have multiple columns here at the Reduce phase